mlir/lib/Dialect/MIOpen/Transforms/AlignTiling.cpp

```diff
@@ -95,94 +96,52 @@ template <typename T> struct MILARewritePattern : public OpRewritePattern<T> {
-    // 1 affine map)
-    auto expr = getAffineConstantExpr(0, ctx);
-    unsigned stride = 1;
-    for (int i = outputDims - 1; i >= 0; --i) {
-      expr =
-          expr + getAffineDimExpr(i, ctx) * getAffineConstantExpr(stride, ctx);
-      strides[i] = b.getIndexAttr(stride);
-      stride *= outputShape[i];
-    }
-    AffineMap outputAffineMap =
-        AffineMap::get(outputDims, 0, ArrayRef<AffineExpr>{expr}, ctx);
-    auto transformedOutputType =
-        MemRefType::get(outputShape, outputElementType, {outputAffineMap});
-
-    llvm::SmallVector<NamedAttribute, 3> transformedNewOutputAttrs;
-    auto transform = b.create<miopen::TransformOp>(
-        loc, transformedOutputType, inp, transformedNewOutputAttrs, true);
-
-    // 2. reduce scope of inp to tile size
-    //      - use subview:
-    //      https://mlir.llvm.org/docs/Dialects/MemRef/#memrefsubview-mlirmemrefsubviewop
-    auto regs = twcopy.getOperand(0);
-    auto regShape = regs.getType().template cast<MemRefType>().getShape();
-    auto regDims = regShape.size();
-    for (uint32_t i = 0; i < regDims; ++i) {
-      Value idx = b.create<IndexCastOp>(loc, twcopy.getOperand(2 + regDims + i),
-                                        b.getIndexType());
-      offsets[i] = idx;
-      sizes[i] = b.getIndexAttr(regShape[i]);
-    }
-
-    return b.create<memref::SubViewOp>(loc, transform, offsets, sizes, strides);
+  Value backtraceToAlloc(Value inp) const {
+    if (auto bt = inp.getDefiningOp<miopen::TransformOp>())
+      return backtraceToAlloc(bt->getOperand(0));
+    if (auto bt = inp.getDefiningOp<memref::ExpandShapeOp>())
+      return backtraceToAlloc(bt->getOperand(0));
+    if (auto bt = inp.getDefiningOp<memref::AllocOp>())
+      return inp;
+    if (auto bt = inp.getDefiningOp<miopen::GpuAllocOp>())
+      return inp;
+    return Value();
   }
 
-  Value traceToThreadwiseCopy(Value inp,
-                              SmallVector<Value, 5> &transforms) const {
-    Value ret;
-    Value laReshape;
-    // 1. get reader (linagl.reshape), return result
-    int cnt = 0;
-    for (auto &use : inp.getUses()) {
-      if (auto op = dyn_cast<linalg::GenericOp>(use.getOwner())) {
-        // reader
-      } else if (!laReshape) {
-        laReshape = getOpResult<memref::ExpandShapeOp>(use);
-      }
-      cnt++;
-    }
-    if (cnt > 2) {
-      return ret;
-    }
-    if (laReshape) {
-      transforms.push_back(laReshape);
-      // 2. get reader (miopen.transform), return result
-      Value miTransform = backtrace<miopen::TransformOp>(laReshape);
-      transforms.push_back(miTransform);
-      Value miTransform2 = backtrace<miopen::TransformOp>(miTransform);
-      transforms.push_back(miTransform2);
-      auto twcopy = dyn_cast<miopen::ThreadwiseCopyOp>(
-          miTransform2.use_begin()->getOwner());
-      if (twcopy)
-        ret = inp;
+  Value makeThreadwiseCopy(PatternRewriter &b, Operation *miTWCopy,
```

Comment:
client logic using this function is making an incorrect assumption that the source memref is smaller than the destination memref, while actually the source memref is much larger than the destination memref.

Therefore the loop bounds becomes too large and the compiler is busy emitting the fully unrolled loop.

Let me think if there's a way to hint `threadwise_copy` that the loop bounds be determined by the destination memref.

Reply:
According to comments in `computeSliceLengths`:

```
  // Order to decide the slice lengths:
  // - bound attribute.
  // - domain attribute from the source in case both source and dest has affine
  //   transformations.
  // - shape of the dest in case only the source has affine transformations.
  // - shape of the source in case the source has no affine transfromations.
```

I think we just need to assign `bound` attribute.

Reply:
NOT the dimensions when we `miopen.alloc` it, but the dimensions after we `miopen.transform` it.